### ISSUE TITLE
[<Ray Compiled Graphs>] A MPMD Graph controller focus on N-M data transfer in complex task

### ISSUE THREAD

**Issue by MoFHeka** [ray-project/ray#48556]
### **Concept introduction**
The fact that SPMD has no scheduling overhead gives it the best performance, but it is often not easy enough to develop complex training tasks. For example, LLM reinforcement learning and recommendation system. At present, the most famous MPMD projects include [Alpa](https://github.com/alpa-projects/alpa) and Google's Pathways (not open source). 
The picture below is an interpretation of the Pathways paper:
![7555dfcda8b169a174f22a6347171296](https://github.com/user-attachments/assets/700e5698-31e2-40ec-8838-fbb89d877906)

In the first step, when the user wants to run a traced program (MPMD) composed of multiple computations (SPMDs), use the Pathways system by calling the Client library.
 
In step 2, the client assigns virtual devices to the computation that has not been run before, and registers the computation with the Resource Manager, that is, obtains the physical devices corresponding to the virtual devices.
 
In step ③, the client triggers the back-end server to compile the computations: First, the MLIR-based dialect constructs a device-agnostic intermediate representation (IR), which is progressively lowered into a low-level representation that senses the physical location of the device through a series of standard passes. It is worth noting that this low-level representation already takes into account network connections between physical devices and data routing operations between different computations. If the physical device corresponding to the virtual device of the program does not change, the generated low-level representation can be reused very quickly; if the Resource Manager changes the virtual device of a program, it needs to be recompiled.
 
Personally, these three steps are referred to as the front-end of Pathways, which is basically a conventional approach and is easier to understand, but it has not yet entered the core of Pathways.
 
In step ④, the Client converts the compiled underlying representation into a dataflow graph that can be recognized by PLAQUE and sent to the runtime system for execution. The Pathways runtime system is missing many details, especially relying on a vague closed-source system PLAQUE, which is a closed-source data flow execution engine used internally by Google.
 
After the distribution is represented as a data flow diagram, it is hosted to the PLAQUE for execution, and the distributed coordination mechanism is its core. PLAQUE consists of two long-running daemons, scheduler and executor, which can communicate with each other (arrows ⑤⑥⑦) to complete the distributed coordination (coordination), which is the communication of the control plane. The mechanism of the control plane is discussed in detail below.
 
The executor uses traditional deep leaming frameworks such as TensorFlow, PyTorch, or JAX (arrow⑧) to complete deep learning computations or data plane communication (mainly cluster communication) represented by arrow⑨. You should be familiar with traditional deep leaming frameworks, and this article will not introduce more.
 
It is worth noting that the communication between the data plane and the control plane may use different transmission media. The former is the blue arrow indicating ICI with higher bandwidth, and the latter is the green arrow indicating DCN with lower bandwidth.

### **Implementation details**
As mentioned above, in fact, different SPMDs programs often have different degrees of data parallelism, so the transmission pipeline(arrow⑨) between them is often N to M. As shown in the following picture:
[SPMD间的传输情况.pdf](https://github.com/user-attachments/files/17626285/SPMD.pdf)

Given the complexity of the P2P push data pipeline, it may also be necessary to set up an abstract structure of **Receiver-Buffer_pool-Computer-Sender** at each node, in which the **Sender** can receive flow control information from its next **Receiver**. As shown in the figure, all data transfer node should inherit this contract:
[抽象数据节点.pdf](https://github.com/user-attachments/files/17626378/default.pdf)
@kevin8542

### Use case

https://github.com/alpa-projects/alpa
Alpa has done a good demonstration in large model training. 
In the recommended system scenario, for example, Ray Data can be read and pushed directly to the GPU/CPU heterogeneous clusters for data pre-processing, and then pushed to the GPU clusters for synchronization training.

**Comment by kevin85421** [comment]
Thank you!

**Comment by kevin85421** [comment]
Thank you for writing the detailed description! I have a few questions:

* In the `Receiver-Buffer_pool-Computer-Sender` diagram, what roles do `flow control message` and `flow control #1` play? How do they interact with `Computer #0`, `Sender #0`, and `Receiver #1`?

* Ray Compiled Graphs currently has the model similar to `Receiver-Buffer_pool-Computer-Sender`. A DAGNode consists of 3 operations, read, compute, and write ([code](https://github.com/ray-project/ray/blob/7aa3cdbe52ce6a0caae3df7a51f4025e056f1d75/python/ray/dag/dag_node_operation.py#L610)). What's your request here?

Thanks!

**Comment by MoFHeka** [comment]
In my understanding, the flow control mechanism is to deal with backpressure feature while maximizing bandwidth efficiency during data-push mode. The Bottleneck Bandwidth and Round-trip propagation time (BBR) algorithm that adjusts traffic according to state machine ACK and bandwidth may be a suitable choice? The way to adjust the traffic can be to change buffer size or send less or compute less or ack to upstream nodes and fetch less from upstream. 
'How do they interact with Computer #0, Sender #0, and Receiver #1?'. I don't have that detail part in mind, it might depend on the specific task.

The 'read, compute, and write' schema in Ray Compiled Graphs is good. But many times the rates of producers and consumers do not match exactly. For example, if the consumer computes more slowly than the producer, that's fine, only one copy of buffer is needed, and there's no need for complicated flow control. 
However, if the consumer computes faster than the producer, then often **multiple producers** are needed, and we do not know the rate ratio between the producer and the consumer in advance, so a flow control algorithm is needed to control maybe-too-many producers(numbers or rates) and adjust a appropriate buffer size.

**Comment by stephanie-wang** [comment]
Thanks for the writeup, @MoFHeka! Can you clarify what exactly you are asking for? Ray graphs are already MPMD and support N-M data transfer, although for sure right now the implementation is pretty naive (see https://github.com/ray-project/ray/issues/47938). [Backpressure](https://github.com/ray-project/ray/issues/43826) is also supported.

If you are asking about cases where producers and consumers execute totally asynchronously, then that is a pretty significant departure from the current system, so it's unclear we would support that anytime soon. You can think of compiled graphs like physical DAGs, so that means we don't do typical logical DAG transforms like automatically sharding a task. The hypothesis is that logical DAG transforms can be done less frequently, either at the beginning of execution or in response to some load change.

For training cases, even with Alpa, this is sufficient as long as you're using synchronous training. Inference is trickier if you want to do dynamic routing or scaling, but assuming that you're okay not changing the physical topology too often, then the application layer should be to handle such transforms by recompiling a new physical DAG(s).

**Comment by MoFHeka** [comment]
@stephanie-wang Thank you for your reply, which would look something like this if written in code:
```python
with InputNode() as inp:
  # no need driver transfer any data except control messages 
  t_a = a.foo
  t_b = b.foo
  t = t.with_type_hint(TorchTensorType(transport="nccl"))
  t_b = b.foo.bind(t)
  t_c = c.foo.bind(t)
  dag = a.update(t, t_b, t_c)
```

**Comment by stephanie-wang** [comment]
Ah I see, thanks! So what you are looking for is a higher-level interface that can express dependencies between device meshes, not just single-device actors, is that right? I think this makes a lot of sense, although currently we don't have bandwidth to support it yet. Welcome any contributions or RFCs on it though!

**Comment by MoFHeka** [comment]
Any progress?
